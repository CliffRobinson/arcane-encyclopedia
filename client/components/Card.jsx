import React from "react";
import { Tooltip } from "react-tippy";

const Card = (props) => {
    let cardSize = "small";
    let hoverCardSize = "normal";

    switch (props.card.layout) {
    case "normal":
    case "saga":
    case "split":
        return (
            <React.Fragment> 
                <Tooltip
                    trigger='mouseenter'
                    html={<img src={props.card.image_uris[hoverCardSize]}></img>}
                >   
                    
                    <img src={props.card.image_uris[cardSize]}></img>
                    <button onClick={()=> console.log(props.card)}> Log this card </button>
                </Tooltip>
            </React.Fragment>
        );

    case "transform":
        return props.card.card_faces.map((face, i) => {
            return (
                <React.Fragment>
                    <Tooltip
                        trigger='mousenter'
                        html = {<img src={face.image_uris[hoverCardSize]}></img>}
                    >
                        
                        <img src={face.image_uris[cardSize]}></img>
                    </Tooltip>
                    <button onClick={()=> console.log(props.card)}> Log this card </button>
                </React.Fragment>
            );            
        });
    default:
        return (
            <div style={{maxWidth:"146 px", border:"1px solid black"}}>
                
                The card {props.card.name} has a weird format and we can't display it. 
                <button onClick={()=> console.log(props.card)}> Log this card </button>
            </div>
        );
        
    }
};

export default Card;