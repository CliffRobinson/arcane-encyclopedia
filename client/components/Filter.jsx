import React from "react";

class Filter extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            function:"customTextFilter",
            key:"type_line",
            value:"",
            exact:false,
        };
        this.handleSelectChange = this.handleSelectChange.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
    }

    handleSelectChange(e, props){
        //console.log("Adding", e.target.value, "to filter", props.i);
        const funcKey = e.target.value.split("--");
        const obj = {
            function:funcKey[0],
            key:funcKey[1],
            value:this.state.value,
            exact:(false || funcKey[2])
        };
        this.setState(obj);
        props.addFilter(obj, props.i);
    }
    
    handleInputChange(e, props){
        //console.log("Adding", e.target.value, "to filter", props.i);
        const value = this.state.function.includes("number") ? parseFloat(e.target.value) : e.target.value;
        const objNixilis = {
            function:this.state.function,
            key:this.state.key,
            value:value,
            exact:this.state.exact
        };
        this.setState(objNixilis);
        props.addFilter(objNixilis, props.i);
    }

    render() {
        return <span className={"filter"}>
            <select className={`select ${this.props.size}`} onChange={(e) => this.handleSelectChange(e, this.props)}>
                <option value="customTextFilter--type_line"> Type Line contains </option>
                <option value="customTextFilter--name"> Name contains</option>
                <option value="customTextFilter--oracle_text"> Oracle Text Contains</option>
                <option value="customTextFilter--rarity--true"> Rarity is</option>
                <option value="numberLessFilter--cmc"> CMC less than </option>
                <option value="numberEqualsFilter--cmc"> CMC equals </option>
                <option value="numberMoreFilter--cmc"> CMC more than </option>
                <option value="numberLessFilter--usd"> Price less than </option>
                <option value="numberMoreFilter--usd"> Price more than </option>
            </select>
            <input className={`${this.props.size}`} type="text" onChange={(e) => this.handleInputChange(e, this.props)}>
    
            </input>
        </span>;
    }
}

export default Filter;