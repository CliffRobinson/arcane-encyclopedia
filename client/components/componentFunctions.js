import request from "superagent";
import {addCards, clearCards } from "../actions/cards";

export function filterAllCards(cards, mana, onlyTricks, excludeLands, sort, customFilters) {
    //console.log("filtering all");
    let output = cards.filter((card) => canCastCardWithMana(card, mana));
    if (excludeLands) {
        output = filterLands(output);
    }
    if (onlyTricks) {
        output = filterForTricks(output);
    }
    if (customFilters) {
        customFilters.map((filter) => {
            output = filterFuncs[filter.function](output, filter.key, filter.value, filter.exact);
        });
    }
    if(sort) {
        output = output.sort(sort);
    }



    return output;
}

export function canCastCardWithMana(card, mana) {
    if (card.cmc > mana.total) {
        return false;
    } else {
        let cardCost = translateMana(card);
        for (let colour in cardCost) {
            if (cardCost[colour] > mana[colour]) return false;
        }
        return true;
    }
}

export function translateMana(card) {
    let manaCost = {
        w: 0,
        u: 0,
        b: 0,
        r: 0,
        g: 0,
        c: 0,
    };
    let stringMana;
    if (card.card_faces) {
        stringMana = card.card_faces[0].mana_cost;
    } else {
        stringMana = card.mana_cost;
    }
    const arrayMana = stringMana.split("");
    arrayMana.map((letter) => {
        if (letter == "W") manaCost.w++;
        if (letter == "U") manaCost.u++;
        if (letter == "B") manaCost.b++;
        if (letter == "R") manaCost.r++;
        if (letter == "G") manaCost.g++;
        if (letter == "C") manaCost.c++;
    });
    return manaCost;
}

export function filterLands(cards) {
    return cards.filter(callbackToFilterLands);
} 

export function callbackToFilterLands(card) {
    switch (card.layout) {
    case "transform":
        return !card.card_faces[0].type_line.includes("Land");
    default:
        return !card.type_line.includes("Land");
    }
}//This might have issues with transform lands. This assumes that only front face determines if a card is a land (IE Westvale Abbey is, Search for Azcanta is not), this might need changing for future sets. 

export function filterForTricks(cards) {

    return cards.filter(callbackToFilterForTricks);
}

export function callbackToFilterForTricks(card) {
    let oText = "";
    switch (card.layout) {
    case "transform":
        if (card.card_faces[0].oracle_text) {
            oText = card.card_faces[0].oracle_text.toLowerCase();
        }
        return (card.card_faces[0].type_line.includes("Instant") || oText.includes("flash"));
    default:
        if (card.oracle_text) {
            oText = card.oracle_text.toLowerCase();
        }
        return (card.type_line.includes("Instant") || oText.includes("flash"));
    }
}

// export function createQuery(){
//     this.props.dispatch(clearCards());
//     let queryString = `https://api.scryfall.com/cards/search?q=${this.props.format}`;
//     getCardsFromScryfall.call(this,queryString);
// }

// export function getCardsFromScryfall(queryString) {
    
//     request.get(queryString)
//         .then( (res) =>{
//             this.props.dispatch(  addCards(res.body.data)  );

//             if (res.body.has_more)  {
//                 getCardsFromScryfall(res.body.next_page);
//             }
//         });
// }

function compareName (cardA, cardiB) {
    if (cardA.name < cardiB.name){
        return -1;
    } else if (cardA.name > cardiB.name) {
        return 1;
    } else {
        console.log("Sort function found 2 cards with the same name. This should not be possible, all MTG cards have unique names.");
        return 0;
    }
} //Cards are alpha sorted by default, so this function is only used as a tiebreaker, as no two differnt cards should have the same name.

function getColorSortIndex(card) {
    let cardColors;

    switch (card.layout){
    case "transform":
        cardColors = card.card_faces[0].colors;
        break;
    default:
        cardColors = card.colors;
    }

    if (cardColors.length == 0){
        return 7;
    } else if (cardColors.length > 1) {
        return 6;
    } else switch(cardColors[0]){
    case "W":
        return 1;
    case "U":
        return 2;
    case "B":
        return 3;
    case "R":
        return 4;
    case "G":
        return 5;
    default:
        console.log(`${card.name} has diamond mana, or something is broken.`);
        return 7;
    }
}

export function getRaritySortIndex(card) {
    switch (card.rarity) {
    case "common":
        return 1;
    case "uncommon":
        return 2;
    case "rare":
        return 3;
    case "mythic":
        return 4;
    default:
        console.log(`${card.name} has invalid rarity.`);
        return 5;
    }
}

export const sortFunctions = {
    compareName: compareName,
    compareCMC: function (cardA, cardB) {
        let diff = cardA.cmc - cardB.cmc;
        if (diff == 0){
            return compareName(cardA, cardB);
        } else return diff;
    },
    comparePrice: function (cardA, cardB) {
        let diff = cardB.usd - cardA.usd;
        if (diff == 0){
            return compareName(cardA, cardB);
        } else return diff;
    },
    compareColor: function (cardA, cardB) {        
        let diff = getColorSortIndex(cardA) - getColorSortIndex(cardB);
        if (diff == 0){
            return compareName(cardA, cardB);
        } else return diff;
    },
    compareCollector: function (cardA, cardB) {        
        let diff = cardA.collector_number - cardB.collector_number;
        if (diff == 0){
            return compareName(cardA, cardB);
        } else return diff;
    },
    compareRarity: function (cardA, cardB) {
        let diff = getRaritySortIndex(cardA) - getRaritySortIndex(cardB);
        if (diff == 0){
            return compareName(cardA, cardB);
        } else return diff;
    }
};

export function customTextFilter(cards, key, text, exact){
    let output = cards;
    output = output.filter((card) => {

        if (exact) {
            if(card[key] || card[key] == 0) {
                return card[key].toLowerCase() == (text.toLowerCase());
            } else if (card.card_faces){
                return card.card_faces[0][key].toLowerCase() == (text.toLowerCase()) || card.card_faces[1][key].toLowerCase() == (text.toLowerCase());
            } else {
                return false;
            }
        } else {
            if(card[key] || card[key] == 0) {
                return card[key].toLowerCase().includes(text.toLowerCase());
            } else if (card.card_faces){
                return card.card_faces[0][key].toLowerCase().includes(text.toLowerCase()) ||card.card_faces[1][key].toLowerCase().includes(text.toLowerCase());
            } else {
                console.log(`${card.name} has no oracle text or does not work with this filter criteria`);
                return false;
            }
        }



    });
    return output;
}

export function numberLessFilter(cards, key, num) {
    return customNumberFilter(cards, key, num, "less");
}

export function numberEqualFilter(cards, key, num) {
    return customNumberFilter(cards, key, num, "equal");
}

export function numberMoreFilter(cards, key, num) {
    return customNumberFilter(cards, key, num, "more");
}

export function customNumberFilter(cards, key, num, comp) {
    let output = cards;
    const compFuncs = {
        less: (a, b) => a < b,
        equal: (a, b) => a == b,
        more: (a, b) => a > b
    };

    return output.filter((card) => {
        if(card[key] || card[key] == 0) {
            return compFuncs[comp](card[key], num);
        } else if (card.card_faces){
            return compFuncs[comp](card.card_faces[0][key], num) || compFuncs[comp](card.card_faces[1][key], num);
        } else {
            console.log(`${card.name} does not work with this filter criteria`);
            return false;
        }
    });
}

export const filterFuncs = {
    customTextFilter,
    numberLessFilter,
    numberEqualFilter,
    numberMoreFilter
};

export function mapManaToProps(state){
    return {
        cards:state.cards, 
        mana:{
            w: state.w,
            u: state.u,
            b: state.b,
            r: state.r,
            g: state.g,
            c: state.c,
            total: (state.w+state.u+state.b+state.r+state.g+state.c) 
        },
        onlyTricks:state.onlyTricks,
        filterLands:state.filterLands,
        sort: state.sort,
        customFilters: state.customFilters
    };
}

