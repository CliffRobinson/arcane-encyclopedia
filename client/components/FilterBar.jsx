import React from "react";
import { connect } from "react-redux";

import Filter from "./Filter";
import { updateFilters, clearFilters } from "../actions/customFilters";


export class FilterBar extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            filters: []
        };
        this.addFilter = this.addFilter.bind(this);
        this.modifyNum = this.modifyNum.bind(this);
    }

    addFilter(filter, i) {
        let newFilters = this.state.filters.slice();
        newFilters[i] = filter;
        this.setState({
            filters: newFilters
        });
    }

    modifyNum(addNotSubtract) {
        
        const newFilters = (addNotSubtract)
            ? [...this.state.filters, null]
            : this.state.filters.slice(0, -1);
        this.setState({
            filters: newFilters
        });
    }

    render() {
        const { filters } = this.state;
        const plural = (this.props.customFilters.length > 1) ? "filters" : "filter";
        const size = "is-small";
        return (
            <div className="filterBar ">
                <button className={`button ${size}`} onClick={() => this.modifyNum(true)} >Add a Filter </button>
                {filters.length > 0 && <button className={`button ${size}`} onClick={() => this.modifyNum(false)}>Remove a Filter</button>}
                {filters.map((e, i) => <Filter size={size} addFilter={this.addFilter} key={i} i={i} />)}
                {filters.length > 0 && <button className={`button ${size}`} onClick={() => this.props.dispatch(updateFilters(this.state.filters))}> Apply {plural}</button>}
                {this.props.customFilters.length > 0 && <button className={`button ${size}`} onClick={() => this.props.dispatch(clearFilters())}> Clear {plural} </button>}
            </div>
        );
    }

}

export const mapCateToProps = (state) => state;

export default connect(mapCateToProps)(FilterBar);