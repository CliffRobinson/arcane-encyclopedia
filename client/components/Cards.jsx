import React from "react";
import { connect } from "react-redux";

import Card from "./Card";
import { filterAllCards, sortFunctions, mapManaToProps } from "./componentFunctions";

export class Cards extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        if (this.props.cards.length != 0) {
            const cardsToDisplay = filterAllCards(this.props.cards, this.props.mana, this.props.onlyTricks, this.props.filterLands, sortFunctions[this.props.sort], this.props.customFilters, this.props.foretell);
            return (
                <div className="cards" >
                    <p> There are {this.props.cards.length} cards in the format </p>
                    <p> Displaying {cardsToDisplay.length} cards </p>
                    {this.props.cards.length && cardsToDisplay.map((card, i) => <Card key={i} card={card} index={i}/>)}
                </div>
            );
        } else {
            return <div className="cards" >
                <p className="title" style={{color:"white"}} > Welcome to the Arcane Encyclopedia! </p>
                <p> This is a tool for seeing what cards your opponent might play against you in Magic: the Gathering. </p>
                <p> To use the app, start by choosing a format in the selector at the top left.</p>
                <p> Use the buttons on the right to select how much mana is available to your opponent. </p>
                <p> The cards they can play with that mana will be shown here. </p>
                <p> If you want to see only instant-speed cards, check the box marked 'Only show tricks' </p>
            </div>;
        }
    }

}

export default connect(mapManaToProps)(Cards);