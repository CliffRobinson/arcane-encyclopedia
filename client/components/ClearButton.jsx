import React from 'react';
import {connect} from 'react-redux';

import {clearMana} from "../actions/mana";
import {mapManaToProps} from "./componentFunctions";


export class ClearButton extends React.Component {
    constructor(props) {
        super(props);
        this.handleClick = this.handleClick.bind(this);
    }

    handleClick() {
        this.props.dispatch(clearMana());
        this.props.clearSelectorMana();
    }

    render() {
        return (
            <div>
                <button className="button is-small" onClick={ this.handleClick }> Clear Mana (component)</button>
            </div>);
    }
}

export default connect(mapManaToProps)(ClearButton);