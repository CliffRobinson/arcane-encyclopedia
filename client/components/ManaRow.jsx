import React from "react";
import { connect } from "react-redux";
import { mapManaToProps } from "./componentFunctions";

export class ManaRow extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            num:0, 
        };
        this.add = this.add.bind(this);
        this.sub = this.sub.bind(this);
    }

    render() {

        if(this.props.showHybrid || this.state.num != 0){        
            return (
                <div className="manaRow">
                    <img className="manaButton" src="/images/mtga-button-transparent.png" onClick={ this.add } />
                    <span  className="manaContainer">
                        <i className={this.props.icon}></i>
                        <span className="manaText">{this.state.num != 0 && this.state.num }</span> 
                    </span>
                    <img className="manaButton" src="/images/mtga-button-minus.png" onClick={ this.sub } />
                </div>
            );
        } else {
            return null;
        }
    }

    add() {
        this.setState({
            num: this.state.num + 1
        });
        this.props.add();
    }

    sub() {
        if (this.state.num > 0) {
            this.setState({
                num: this.state.num - 1
            });
            this.props.sub();
        }
    }



}

export default connect(mapManaToProps)(ManaRow);