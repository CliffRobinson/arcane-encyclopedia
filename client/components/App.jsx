import React from "react";
import Cards from "./Cards";
import ManaSelector from "./ManaSelector";
import SearchSelector from "./SearchSelector";
import FilterBar from "./FilterBar";


const App = () => {
    return (
        <div className="App" >
            <div className="topTwo">
                <SearchSelector />
                <FilterBar />
            </div>

            <div className="flexContain">
                <div className="main">
                    <Cards />
                </div>
                <div className="sideColumn">
                    <ManaSelector />
                </div>
                
            </div>
        </div>
    );
};

export default App;

