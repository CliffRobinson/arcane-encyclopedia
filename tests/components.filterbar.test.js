import React from 'react';
import {shallow, mount} from 'enzyme';

import {FilterBar} from "../client/components/FilterBar";

const fakeFilters = [
    {
        function: "customTextFilter",
        key: "type_line",
        value: "human",
        exact: false,
    },
    {
        function: "customTextFilter",
        key: "rarity",
        value: "uncommon",
        exact: "true",
    }
];

test("Filterbar matches snapshot when no filters are present", () => {
    expect(
        toJson(shallow(<FilterBar customFilters={[]}/>))
    ).toMatchSnapshot();
});

test("Filterbar matches snapshot when filters are present", () => {
    expect(
        toJson(shallow(<FilterBar customFilters={fakeFilters}/>))
    ).toMatchSnapshot();
});

test("first button calls modifyNum to add a filter", () => {
    //Arrange
    const testFilterBar = shallow(<FilterBar customFilters={[]}/>);
    const spy = jest.spyOn(testFilterBar.instance(), "modifyNum");
    const buttons = testFilterBar.find("button");
    //Act
    buttons.at(0).props().onClick();
    //Assert
    expect(spy.mock.calls[0][0]).toBe(true);
});