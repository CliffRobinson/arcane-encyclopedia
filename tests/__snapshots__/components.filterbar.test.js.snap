// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filterbar matches snapshot when filters are present 1`] = `
<div
  className="filterBar "
>
  <button
    className="button is-small"
    onClick={[Function]}
  >
    Add a Filter 
  </button>
  <button
    className="button is-small"
    onClick={[Function]}
  >
     Clear 
    filters
     
  </button>
</div>
`;

exports[`Filterbar matches snapshot when no filters are present 1`] = `
<div
  className="filterBar "
>
  <button
    className="button is-small"
    onClick={[Function]}
  >
    Add a Filter 
  </button>
</div>
`;
